runtime/near-vm-logic/src/logic.rs

```diff
@@ -2786,6 +2786,56 @@ impl<'a> VMLogic<'a> {
         }
         Ok(())
     }
+
+    /// Verify an ED25519 signature given a message and a public key.
+    /// # Returns
+    /// - 1 meaning the boolean expression true to encode that the signature was properly verified
+    /// - 0 meaning the boolean expression false to encode that the signature failed to be verified
+    ///
+    /// # Cost
+    ///
+    /// Each input can either be in memory or in a register. Set the length of the input to `u64::MAX`
+    /// to declare that the input is a register number and not a pointer.
+    /// Each input has a gas cost input_cost(num_bytes) that depends on whether it is from memory
+    /// or from a register. It is either read_memory_base + num_bytes * read_memory_byte in the
+    /// former case or read_register_base + num_bytes * read_register_byte in the latter. This function
+    /// is labeled as `input_cost` below.
+    ///
+    /// `input_cost(num_bytes_signature) + input_cost(num_bytes_message) + input_cost(num_bytes_public_key) +
+    ///  ed25519_verify_base + ed25519_verify_byte * (num_bytes_signature + num_bytes_message)`
+    #[cfg(feature = "protocol_feature_ed25519_verify")]
+    pub fn ed25519_verify(
+        &mut self,
+        sig_len: u64,
+        sig_ptr: u64,
+        msg_len: u64,
+        msg_ptr: u64,
+        pub_key_len: u64,
+        pub_key_ptr: u64,
+    ) -> Result<u64> {
+        use ed25519_dalek::{PublicKey, Signature, Verifier};
+
+        let msg = self.get_vec_from_memory_or_register(msg_ptr, msg_len)?;
+        let signature_array = self.get_vec_from_memory_or_register(sig_ptr, sig_len)?;
+        self.gas_counter.pay_base(ed25519_verify_base)?;
+        let num_bytes = msg
+            .len()
+            .checked_add(signature_array.len())
```

Comment:
@jakmeier conflicting suggestion here with regards to the byte cost:
```
    ///  ed25519_verify_base + ed25519_verify_byte * (num_bytes_signature + num_bytes_message)`
```

Reply:
Right. I thought because we have `sig_len: u64` as a parameter there is somehow a possibility that the signature length can vary now or maybe in the future. But I guess the function name makes it clear that this is only used with `ed25519` which fully defines the signature and public key length.

I would say don't charge per-byte costs for PK and signature. Only for the message. And return `Ed25519VerifyInvalidInput` if the input length is not equal to the corresponding constant.

Unfortunately, we still need both `sig_len: u64` and `pub_key_len: u64` as a parameters to decide on whether to read from a register or from memory. I would love to see something more elegant but cannot think of a simple solution. It's okay to keep them IMO.